dlt/destinations/impl/sink/sink.py

```diff
@@ -0,0 +1,170 @@
+from abc import ABC, abstractmethod
+from types import TracebackType
+from typing import ClassVar, Dict, Optional, Type, Iterable, Iterable
+
+from dlt.destinations.job_impl import EmptyLoadJob
+from dlt.common.typing import TDataItems
+from dlt.common import json
+from dlt.common.pipeline import load_package_state, commit_load_package_state
+
+from dlt.common.schema import Schema, TTableSchema, TSchemaTables
+from dlt.common.schema.typing import TTableSchema
+from dlt.common.storages import FileStorage
+from dlt.common.destination import DestinationCapabilitiesContext
+from dlt.common.destination.reference import (
+    TLoadJobState,
+    LoadJob,
+    JobClientBase,
+)
+
+from dlt.destinations.impl.sink import capabilities
+from dlt.destinations.impl.sink.configuration import SinkClientConfiguration, TSinkCallable
+
+
+class SinkLoadJob(LoadJob, ABC):
+    def __init__(
+        self,
+        table: TTableSchema,
+        file_path: str,
+        config: SinkClientConfiguration,
+        schema: Schema,
+        load_package_state: Dict[str, int],
+    ) -> None:
+        super().__init__(FileStorage.get_file_name_from_file_path(file_path))
+        self._file_path = file_path
+        self._config = config
+        self._table = table
+        self._schema = schema
+
+        self._state: TLoadJobState = "running"
+        self._storage_id = f"{self._parsed_file_name.table_name}.{self._parsed_file_name.file_id}"
+        try:
+            current_index = load_package_state.get(self._storage_id, 0)
+            for batch in self.run(current_index):
+                self.call_callable_with_items(batch)
+                current_index += len(batch)
+                load_package_state[self._storage_id] = current_index
+
+            self._state = "completed"
+        except Exception as e:
+            self._state = "retry"
+            raise e
+        finally:
+            # save progress
+            commit_load_package_state()
+
+    @abstractmethod
+    def run(self, start_index: int) -> Iterable[TDataItems]:
+        pass
+
+    def call_callable_with_items(self, items: TDataItems) -> None:
+        if not items:
+            return
+
+        # coerce items into correct format specified by schema
+        coerced_items: TDataItems = []
+        for item in items:
+            coerced_item, table_update = self._schema.coerce_row(self._table["name"], None, item)
```

Comment:
why we need to do it. this is expensive operation. all rows are coerced in the normalizer?

Reply:
@rudolfix I'd still need an answer on this one.

Reply:
use jsonl-pue file format and prevent using jsonl. `pue` has additional encoding and and a set of functions that recursively restore python objects.